app/controllers/lti_nrps_controller.rb

```diff
@@ -84,11 +84,86 @@ def send_nrps_request
     lcd.last_synced = DateTime.current
     lcd.save
 
-    render json: members.as_json
+    # Update the roster with the retrieved set of members
+    @cuds = parse_members_data(lcd, members.as_json)
+    @sorted_cuds = @cuds.sort_by { |cud| cud[:color] || "z" }.reverse
+    @course = lcd.course
   end
 
 private
 
+  def populate_cud_data(cud)
+    user = cud.user
+    {
+      email: user.email,
+      first_name: user.first_name,
+      last_name: user.last_name,
+      course_number: cud.course_number,
+      lecture: cud.lecture,
+      section: cud.section,
+      school: user.school,
+      major: user.major,
+      year: user.year,
+      grade_policy: cud.grade_policy
+    }
+  end
+
+  def parse_members_data(lcd, members_data)
+    cuds = CourseUserDatum.where(course_id: lcd.course_id, instructor: false,
+                                 course_assistant: false).to_set
+    email_to_cud = {}
+    cuds.each do |cud|
+      email_to_cud[cud.user.email] = cud
+    end
+
+    cud_view = []
+    members_data.each do |user_data|
+      next unless user_data["roles"].include? "http://purl.imsglobal.org/vocab/lis/v2/membership#Learner"
+
+      next if user_data.key?("status") && user_data["status"] != "Active"
+
+      next unless user_data.key?("email") && user_data.key?("given_name") &&
+                  user_data.key?("family_name")
+
+      # Normalize email
+      user_data["email"].downcase!
+
+      cud_data = {}
+      user = User.find_by(email: user_data["email"])
+      if user.nil? || lcd.course.course_user_data.find_by(user_id: user.id).nil?
+        cud_data[:color] = "green"
+        cud_data[:email] = user_data["email"]
+        cud_data[:first_name] = user_data["given_name"]
+        cud_data[:last_name] = user_data["family_name"]
+        unless user.nil?
```

Comment:
Added in order to preserve an existing user's Major, School, and Year information on a new CUD